program/src/processor.rs

```diff
@@ -188,20 +189,177 @@ impl Processor {
         let rent = Rent::get()?;
         check!(
             rent.is_exempt(mango_account_ai.lamports(), size_of::<MangoAccount>()),
-            MangoErrorCode::Default
+            MangoErrorCode::AccountNotRentExempt
         )?;
         check!(owner_ai.is_signer, MangoErrorCode::SignerNecessary)?;
 
-        let _mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        let mango_group = MangoGroup::load_checked(mango_group_ai, program_id)?;
+        check!(
+            mango_group.num_mango_accounts <= mango_group.max_mango_accounts,
+            MangoErrorCode::MaxAccountsReached
+        )?;
 
         let mut mango_account: RefMut<MangoAccount> = MangoAccount::load_mut(mango_account_ai)?;
         check_eq!(&mango_account_ai.owner, &program_id, MangoErrorCode::InvalidOwner)?;
-        check!(!mango_account.meta_data.is_initialized, MangoErrorCode::Default)?;
+        check!(!mango_account.meta_data.is_initialized, MangoErrorCode::InvalidAccountState)?;
 
         mango_account.mango_group = *mango_group_ai.key;
         mango_account.owner = *owner_ai.key;
         mango_account.order_market = [FREE_ORDER_SLOT; MAX_PERP_OPEN_ORDERS];
         mango_account.meta_data = MetaData::new(DataType::MangoAccount, 0, true);
+        mango_account.not_upgradable = true;
+        Ok(())
+    }
+
+    #[inline(never)]
+    fn close_mango_account(program_id: &Pubkey, accounts: &[AccountInfo]) -> MangoResult<()> {
+        const NUM_FIXED: usize = 3;
+        let accounts = array_ref![accounts, 0, NUM_FIXED];
+
+        let [
+            mango_group_ai,     // write
+            mango_account_ai,   // write
+            owner_ai,           // write, signer
+        ] = accounts;
+
+        check_eq!(&mango_account_ai.owner, &program_id, MangoErrorCode::InvalidOwner)?;
+        let mut mango_account =
+            MangoAccount::load_mut_checked(mango_account_ai, program_id, &mango_group_ai.key)?;
+        check_eq!(&mango_account.owner, owner_ai.key, MangoErrorCode::InvalidOwner)?;
+        check!(owner_ai.is_signer, MangoErrorCode::InvalidSignerKey)?;
+        check!(mango_account.meta_data.version > 0, MangoErrorCode::InvalidAccountState)?;
+
+        // Check deposits and borrows are zero
+        for i in 0..MAX_TOKENS {
+            check_eq!(mango_account.deposits[i], ZERO_I80F48, MangoErrorCode::InvalidAccountState)?;
+            check_eq!(mango_account.borrows[i], ZERO_I80F48, MangoErrorCode::InvalidAccountState)?;
+        }
+        // Check no perp positions or orders
+        for perp_account in mango_account.perp_accounts.iter() {
+            check_eq!(perp_account.base_position, 0, MangoErrorCode::InvalidAccountState)?;
+            check_eq!(
+                perp_account.quote_position,
+                ZERO_I80F48,
+                MangoErrorCode::InvalidAccountState
+            )?;
+
+            check!(perp_account.is_liquidatable(), MangoErrorCode::InvalidAccountState)?;
```

Comment:
Why does it need to be liquidatable?

Ah, this actually checks bids/asks/taker_quote/taker_base. Rename function?